### ISSUE TITLE
Adding package to ELPA

### ISSUE THREAD

**Issue by phikal** [minad/consult#75]
Hi,

I was wondering if there were any plans to add this package to GNU ELPA? It seems like it would be a good fit.

**Comment by minad** [comment]
Yes, I thought about it before. I think it is not unreasonable if marginalia, consult etc move closer to upstream since they work well with icomplete too. 

The downside is that it slows down development due to the paperwork. I myself have the papers lying around here and haven't submitted them yet (I should though since I also want to contribute some things to other ELPA projects).

Do you know more about this non-GNU ELPA which is being planned? Is that an acceptable alternative?

**Comment by phikal** [comment]
> The downside is that it slows down development due to the paperwork.

Do you mean because new contributors have to sign the papers? I guess, but in my case it took a week, and just two or three emails, so I can't comment on that.

> Do you know more about this non-GNU ELPA which is being planned? Is that an acceptable alternative?

Maybe this might help: https://git.savannah.gnu.org/cgit/emacs/nongnu.git/tree/README.org#n79

It's still not ready (see https://lists.gnu.org/archive/html/emacs-devel/2020-11/threads.html#00939), but it would also be a good alternative, if the contributors-agreement is an issue. But as I said, the main down-side is that it currently doesn't exist. It could be suggested to add it, but personally, I think that adding it to regular ELPA would be preferable, especially since the project isn't too old and it only gets harder as time goes on.

**Comment by minad** [comment]
@phikal Thank you! I think about it! Regarding ELPA - what do you see as the advantages if the package is added there?

**Comment by phikal** [comment]
Daniel Mendler <notifications@github.com> writes:

> @phikal Thank you! I think about it! Regarding ELPA - what do you see as the advantages if the package is added there?

The main ones are that it is available to all (GNU) Emacs users, and
that it might be merged into the core one day/the core could be adapted
to improve consult.

I, for example, don't use MELPA because it publishes a new version for
each commit, and therefore have to "git clone +
package-install-from-buffer"-it, to test it out. In my case this could
be fixed by adding a tag, because I have MELPA Stable configured, but
even then, it is not something you can depend on (MELPA Stable packages
can depend on MELPA packages, which can break stuff).

When I have a ELPA and a MELPA package, I usually prefer the ELPA one,
because it feels more reliable. It makes sense after all, because
basically the "worlds best Elisp experts" can comment on it, and suggest
idiomatic improvements. But I guess the same will also apply to NonGNU
ELPA.

-- 
	Philip K.

**Comment by minad** [comment]
Thanks!

> The main ones are that it is available to all (GNU) Emacs users

Agree.

> and that it might be merged into the core one day/the core could be adapted
to improve consult.

About this I am unsure - I don't think it is an advantage if more and more things are merged into Emacs core. I would personally prefer if stuff is removed from core and moved to ELPA. From my perspective there is already enough cruft in the Emacs main repository.

> I, for example, don't use MELPA because it publishes a new version for
each commit, and therefore have to "git clone +
package-install-from-buffer"-it, to test it out. In my case this could
be fixed by adding a tag, because I have MELPA Stable configured, but
even then, it is not something you can depend on (MELPA Stable packages
can depend on MELPA packages, which can break stuff).

Are you worried about stability or security? I am sure you are aware of alternative package managers which allow reviewing and version pinning.

> When I have a ELPA and a MELPA package, I usually prefer the ELPA one, because it feels more reliable. 

Makes sense. While MELPA is also reviewed quite a bit these days, at least when packages enter, it still leaves some security holes and has stability issues since it pulls from the main branch.

> It makes sense after all, because basically the "worlds best Elisp experts" can comment on it, and suggest
idiomatic improvements. But I guess the same will also apply to NonGNU
ELPA.

Yes, hopefully the review aspect could be as good in nongnu elpa.

**Comment by phikal** [comment]
> About this I am unsure - I don't think it is an advantage if more and more things are merged into Emacs core. I would personally prefer if stuff is removed from core and moved to ELPA. From my perspective there is already enough cruft in the Emacs main repository.

Actually I agree, it's just a fact that code can be moved more easily between ELPA and the core.

> Are you worried about stability or security?

For me it's mainly convenience and ootb usability. And I'm familiar with alternative package managers, but not too enthusiastic.

**Comment by minad** [comment]
> Actually I agree, it's just a fact that code can be moved more easily between ELPA and the core.

Yes, I agree that this is a good thing!

> For me it's mainly convenience and ootb usability. And I'm familiar with alternative package managers, but not too enthusiastic.

I am also using the normal package.el, but things like straight certainly have its appeal with the fine grained locking - did not yet try it though and also didn't have a need as of now. I usually only install packages I consider very solid such that I am not observing too much breakage.

I will think about the ELPA proposal. Maybe some others want to chime in and give their opinion? Maybe it also makes sense to wait a bit to see where this nongnu ELPA goes.

**Comment by phikal** [comment]
> I will think about the ELPA proposal. Maybe some others want to chime in and give their opinion? Maybe it also makes sense to wait a bit to see where this nongnu ELPA goes.

I'll just ping all contributors, because the suggestion depends on everyone who made a significant contribution: @oantolin, @okamsn, @tomfitzhenry, @clemera

**Comment by clemera** [comment]
I have done the paperwork already and have packages on GNU ELPA myself so my contribution would pass. I also think consult would be nice candidate for non-GNU ELPA but as mentioned it could take some time until it becomes available and AFAIK it is not clear how will be operated exactly.

**Comment by minad** [comment]
@clemera I understand that you would generally be okay with such a move. Is this also your opinion/preference? What advantages/disadvantages do you see?

I think I am looking at too much from the package maintainer's perspective. From there I mostly see the paperwork and no real technical advantages - but from the user perspective, having things on ELPA may make a difference?

**Comment by clemera** [comment]
I'm undecided, it's nice that packages are available without the need for users to adjust their package archives on the other hand it's also nice to keep it easy for users to contribute and consult is a good candidate for potentially many contributions without users needing to be too familiar with its code base as it collects generic commands.

**Comment by minad** [comment]
@clemera Agree, this is pretty much aligned with how I see it. I think many people already use MELPA, so only very few users benefit by not having to make the package archive adjustment.

**Comment by tomfitzhenry** [comment]
> I'll just ping all contributors, because the suggestion depends on everyone who made a significant contribution: @oantolin, @okamsn, @tomfitzhenry, @clemera

I'm happy for my contributions to be copyright assigned to FSF, and am already on FSF's file.

I tthink Consult could benefit from contributions from Emacs core (more potential to improve the minibuffer) and from Github (more contributors). Requiring a license agreement may deter Github contributors.

company-mode and eglot both receive contributions from Github, despite requiring FSF copyright assignment:
* https://github.com/company-mode/company-mode.github.com/blob/1cd76ade5315b692de9be52a23f81ee3fb043abc/index.html#L138-L146 
* https://github.com/joaotavora/eglot/pull/505#issuecomment-664231610

A few options:
1. Try it and see. Start requiring FSF copyright assignment. If this option is chosen, the sooner the better (to avoid https://github.com/jwiegley/use-package/issues/282 ). If this causes a drop off in contributors, then we could revert to not requiring copyright assignment.
2. Wait for NonGNU ELPA, and try to get into that. Will NonGNU ELPA be part of `package-archives` by default?
3. Stay in MELPA.

I lean towards (2), because of the uncertain cost of dealing with paperwork.

**Comment by phikal** [comment]
> more potential to improve the minibuffer

That's another good example.

> company-mode and eglot both receive contributions from Github, despite requiring FSF copyright assignment:

A few more examples:

- [yasnippet](https://github.com/joaotavora/yasnippet/graphs/contributors)
- [avy](https://github.com/abo-abo/avy/graphs/contributors)
- [exwm](https://github.com/ch11ng/exwm/graphs/contributors)
- [ggtags](https://github.com/leoliu/ggtags/graphs/contributors)
- [realgud](https://github.com/realgud/realgud/graphs/contributors)
- [swiper/ivy/counsel](https://github.com/abo-abo/swiper/graphs/contributors)

> Wait for NonGNU ELPA, and try to get into that. Will NonGNU ELPA be part of package-archives by default?

Yes, that's exactly what will make it interesting! In principle, you don't have to wait, [the repo is being worked on](https://git.savannah.gnu.org/cgit/emacs/nongnu.git/refs/), which each package in a branch.

> How much time it'd take maintainers to upload new versions?

In my experience, it takes less than a day, but you can request access to the repos for yourself as a developer, and do it yourself.

> What workflows that maintainer would have to learn/remember (patches via email, I suppose?)

I can't say for sure, [in my case](http://elpa.gnu.org/packages/shell-command+.html) I didn't have to patch anything, just suggested adding it on `emacs-devel` and it was added, because I was the only contributor ([this](https://lists.gnu.org/archive/html/emacs-devel/2020-09/msg01958.html) was the thread, in case anyone is interested). I guess the best thing is to ask the mailing list, they know for sure.

> How much time it'd take contributors to sign FSF forms? 2h effort and 1-2w of back and forth paperwork?

Again, in my case it was maybe 10 minutes in total, over a week. I requested a copy, recieved it, signed it and sent it back. Depending on your hardware it might be more or less effort, and depending on how much stress the FSF people are the responses might take longer. But I don't think it's too much of a problem, unless someone really doesn't want to sign the papers in principle. It just delays the first pull request by a week, but they you can contribute to a lot more packages.

**Comment by okamsn** [comment]
I have already signed the FSF copyright paperwork. My contributions should not pose a problem.

**Comment by minad** [comment]
Okay, it sounds as if many of you are not against such a move to ELPA.

But for me it is a requirement in any case to get access to the repository. Otherwise I would simply feel limited for no good reason.

Furthermore @oantolin's opinion is important, since he is a major contributor and if such a move is made I would suggest moving marginalia too. And maybe embark.

Can nongnu elpa already be used as package archive? It will probably be included in 28 by default?

**Comment by leungbk** [comment]
> Okay, it sounds as if many of you are not against such a move to ELPA.
> 
> But for me it is a requirement in any case to get access to the repository. Otherwise I would simply feel limited for no good reason.

`ivy` seems to get by with updating its Elpa package only on version upgrades, while updating its Melpa package with each new commit regardless of whether or not Oleh chooses to cut a new stable version of `ivy`. So anyone who wants the bleeding-edge version is fine as long as they opt into Melpa.

**Comment by oantolin** [comment]
I realize I haven't chimed in here. I'm not opposed to adding any of Consult, Marginalia or Embark to GNU ELPA. I also don't see any need to do so, or any (real) advantage ---that is, I've read the supposed advantages and remain unconvinced. If @minad decides to try to get Consult or Marginalia added to GNU ELPA I would also try to get Embark in. I have not signed the FSF papers, but am willing to do so, as long as I can figure out the procedure.

**Comment by minad** [comment]
@oantolin I also don't see big benefits of moving to ELPA. But in particular I would hear more about non-gnu ELPA - this sounds more like an option if we want to keep the current development style without getting involved with bureaucracy.
But I have another tiny package of mine which I will could try to get into ELPA to get familiar with the process, instead of publishing it on MELPA. Then I can tell more.

**Comment by protesilaos** [comment]
> I also don't see big benefits of moving to ELPA.

The only potential benefit is that your code qualifies for inclusion in Emacs from a legal perspective.  So if the Emacs maintainers want to add Consult to core, they will just need a final confirmation from your side.

Otherwise opting for GNU ELPA is mostly a convenience for users who do not want to set up MELPA.

Those granted, do bear in mind that the MELPA maintainers discourage the distribution of the same package via two different archives.  I was informed about it recently when it was discovered that my themes are in Emacs, GNU ELPA, and MELPA.

**Comment by minad** [comment]
> The only potential benefit is that your code qualifies for inclusion in Emacs from a legal perspective. So if the Emacs maintainers want to add Consult to core, they will just need a final confirmation from your side.

This is something I wouldn't even want. I would prefer if more stuff is kept outside of emacs. Rather things should be moved from Emacs core to ELPA.

> Otherwise opting for GNU ELPA is mostly a convenience for users who do not want to set up MELPA.

Agree.

> Those granted, do bear in mind that the MELPA maintainers discourage the distribution of the same package via two different archives.

I think some packages use melpa for the bleeding edge version and elpa for the stable version. Might not be in the interest of the melpa maintainers though.

**Comment by protesilaos** [comment]
> This is something I wouldn't even want. I would prefer if more stuff is kept outside of emacs. Rather things should be moved from Emacs core to ELPA.

I think that trend will only be reversed when/if an Emacs tarball can be produced with packages pulled directly from GNU ELPA.  I noticed some allusions to this notion in emacs-devel but nothing concrete exists right now.

For the rest, it really is up to you.  Non-GNU ELPA sounds promising, though it remains a work-in-progress.

**Comment by minad** [comment]
> I think that trend will only be reversed when/if an Emacs tarball can be produced with packages pulled directly from GNU ELPA. I noticed some allusions to this notion in emacs-devel but nothing concrete exists right now.

This sounds like a really good idea! Would be backward compatible and it would introduce a better separation between the components for development and to track dependencies.

**Comment by phikal** [comment]
> This is something I wouldn't even want. I would prefer if more stuff is kept outside of emacs. Rather things should be moved from Emacs core to ELPA.

It might be worth noting, that this doesn't have to mean that the entire package is merged into the core, but only parts of it, or even a single function/user option/etc., depending on the concrete example.

**Comment by minad** [comment]
@phikal Sure, but I am not seeing that. In consult there is only consult-multi-occur which should actually go upstream, and it is short enough to not require the agreement. Marginalia is different, I guess the annotators there could be useful to go upstream as default annotators. But then I also like the marginalia approach of adding annotations afterwards instead of hard coding the annotators with the commands. So once again it is better not to upstream.

**Comment by minad** [comment]
After having interacted with the ELPA maintainers I feel comfortable about this now and I see the advantages in interacting directly with the Emacs maintainers. This way we can contribute to upstream improvements. My Vertico and Corfu packages are already on ELPA. These two were easy since I am the single contributor.

@oantolin How do you feel about this? There has been some discussion of moving Orderless to ELPA. Are you interested in doing that? If yes, I would propose to move Marginalia and Embark too. Marginalia only needs copyright assignment by myself and @oantolin.

-----

Regarding Consult the situation is a bit more complicated:

* [x] @clemera (consult-bookmark,  ...)
* [x] @okamsn (consult-kmacro, ...)
* [x] @tomfitzhenry (consult-flymake)
* [ ] @oantolin (consult-completion-in-region, consult-file-externally, ...)
* [ ] @astoff (consult-isearch, ...)
* [ ] @omar-polo (consult-man)
* [ ] @inigoserna (consult-find, consult-locate)

Then there was code by @s-kostyaev, but this code has been removed (consult-line-from-isearch). Furthermore there have been a few more small contributions which don't need assignment. Then some of the code in Consult is derived from code on the Selectrum wiki contributed by @okamsn and @clemera.

**Comment by minad** [comment]
I should add, @protesilaos has been a great help in the process. He helped pushing changes to ELPA, so there is no real issue in getting changes published quickly. And at some point I may also request ELPA access.

**Comment by astoff** [comment]
I have already signed the FSF copyright papers some time ago.

**Comment by minad** [comment]
Thank you, @astoff!

**Comment by omar-polo** [comment]
I have already signed the FSF copyright papers too.

**Comment by minad** [comment]
@omar-polo Thank you for the quick response too!

**Comment by protesilaos** [comment]
> I should add, @protesilaos has been a great help in the process. He
> helped pushing changes to ELPA, so there is no real issue in getting
> changes published quickly. And at some point I may also request ELPA
> access.

Thanks again for your contributions!  If/when you make this happen, I am
willing to help you sync commits with elpa.git (just did for corfu and
vertico).

I think requesting access to GNU ELPA is the next logical step.  Note
though that this means becoming a member of the "emacs" group, which
includes both emacs.git and elpa.git.  That will be of use once you
start patching completing-read and friends ;)

-- 
Protesilaos Stavrou
protesilaos.com

**Comment by minad** [comment]
> Thanks again for your contributions!  If/when you make this happen, I am
willing to help you sync commits with elpa.git (just did for corfu and
vertico).

Thanks, that would be great!

**Comment by oantolin** [comment]
I haven't signed the copyright assignment papers yet but will start the process soon. Stefan Monnier seems to want orderless in GNU ELPA and I don't see any reason not to do it (I don't really see good reasons to do it either, other than being asked by Stefan, but that reason alone is enough for me).

**Comment by minad** [comment]
@oantolin Thanks!

It seems Stefan cares a lot about ELPA and the packages there. He has made multiple helpful comments to my packages. And he obviously has great insight as former Emacs maintainer and author of minibuffer.el. Then it helps in case one wants to propose/contribute some changes upstream. These are all good reasons for me now. 

The downsides seem less relevant now. There are not that many contributions to Consult, so I don't worry about the assignment barrier. Then the process to push to ELPA is also very smooth, but this is thanks to @protesilaos in my case! And as mentioned, one can also get direct access at some point.

**Comment by minad** [comment]
For anyone watching here - I prepared Marginalia and Consult for the submission to ELPA, which will happen soon.

**Comment by protesilaos** [comment]
On 2021-05-27, 16:00 -0700, Daniel Mendler ***@***.***> wrote:

> For anyone watching here - I prepared Marginalia and Consult for the
> submission to ELPA, which will happen soon.

Congratulations!

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by minad** [comment]
Package has been submitted. Thank you all! :tada:

**Comment by phikal** [comment]
Great. What an interesting development of an issue.
